### PR TITLE
[iam] Add proto definitions for OIDC provider config

### LINKED ISSUE #14951: [sso/oidc] Proto definitions for OIDC ProviderConfig in public-api
part of Epic: Single sign-on (SSO) #7761

* ProviderConfig entity can be a superset of [oidc.ProviderConfig](https://github.com/coreos/go-oidc/blob/fb9e00951dc7a7d035e7664a5df116c3563afbd7/oidc/oidc.go#L137-L164) or [oidc.RelyingParty](https://github.com/zitadel/oidc/blob/356dd89ae4fad3e95ba118ec3ee51f0b42c97448/pkg/client/rp/relying_party.go#L30-L67)
* Derive useful additional attributes from existing AuthProviders, e.g. descriptions, icons, scopes, hints
* Add CRUD service

### PR DESCRIPTION
**Pull request by AlexTugarev**
This PRs introduces protocol definitions for OIDC provider configurations.

These configurations are meant to define instances of identity providers. There are no assumptions included about the entities these providers will be associated with.

The configuration surface of OIDC providers included here are to cover use cases like Google initially. Further extensions will be added incrementally if/as required. 

## Related Issue(s)

Resolves #14951

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/public-api/gitpod/experimental/v1/oidc.proto [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+message CreateOIDCProviderConfigResponse { OIDCProviderConfig config = 1; }
+
+message GetOIDCProviderConfigRequest { string id = 1; }
+message GetOIDCProviderConfigResponse { OIDCProviderConfig config = 1; }
+
+message ListOIDCProviderConfigsRequest {
+  // Optional.
+  bool admin = 1;
```
Comment (easyCZ) [inline components/public-api/gitpod/experimental/v1/oidc.proto:125]: What does this mean?

File: components/public-api/gitpod/experimental/v1/oidc.proto [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  // ID Token to encode the Claims in a JWT. The algorithm RS256 MUST be
+  // included. Optional.
+  repeated string id_token_signing_alg_values_supported = 21;
+
+  // Time when the config was created.
+  // Read-only.
+  google.protobuf.Timestamp creation_time = 22;
+}
+
+service OIDCProviderConfigService {
```
Comment (easyCZ) [inline components/public-api/gitpod/experimental/v1/oidc.proto:87]: Could we simply this to `OIDCService`, alternatively, `OIDCSProviderService`? Config is effectively what we're putting in the resource so I believe we don't need it in the service title.

We can then also simplify the RPCs to drop the `Config` part

File: components/public-api/gitpod/experimental/v1/oidc.proto (hunk 1 above)
Comment (AlexTugarev) [reply components/public-api/gitpod/experimental/v1/oidc.proto:125]: The intention is to select the configs we're interested in. How would you express if you are interested in seeing admin-level configs? 

I would also be fine starting without a selector for now and implicitly target admin-level. This is somewhat linked to the organization modeling.

Comment (easyCZ) [reply components/public-api/gitpod/experimental/v1/oidc.proto:125]: I'd say let's start with "list all, paginated" which will in effect mean list all the things I have access to. Once see the pattern we see how to structure these as "filter options"

File: components/public-api/gitpod/experimental/v1/oidc.proto (hunk 2 above)
Comment (AlexTugarev) [reply components/public-api/gitpod/experimental/v1/oidc.proto:87]: Thought about that.

Let's have a look at an example which shows the issue with simplifying the name. 

The *OIDC provider* (IdP) may be Google
   * the metadata for this provider are declared https://accounts.google.com/.well-known/openid-configuration
   * that provider supports multiple clients, a single Gitpod installation may have configured several clients at the same time
   * (in fact) the proper term would be "OIDC client configuration"
   * leaving out the configuration removes the necessary additional dimension

File: components/public-api/gitpod/experimental/v1/oidc.proto [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      returns (DeleteOIDCProviderConfigResponse) {};
+}
+
+message CreateOIDCProviderConfigRequest {
+  OIDCProviderConfig config = 1;
+
+  // Optional.
+  bool use_discovery = 2;
+  // Optional.
+  bool is_oauth_only = 3;
```
Comment (easyCZ) [inline components/public-api/gitpod/experimental/v1/oidc.proto:115]:
```suggestion
  bool oauth_only = 3;
```
The boolean type already indicates "the questions you can ask against this field". See https://google.aip.dev/140

File: components/public-api/gitpod/experimental/v1/oidc.proto (hunk 2 above)
Comment (easyCZ) [reply components/public-api/gitpod/experimental/v1/oidc.proto:87]: So the Resource is OIDC Client Configuration, so the underlying object we store (and type here can remain `OIDCClientConfiguration` but the service doesn't need encode the full resource. The service, and the RPC for a unique identifier. So together, we can have

`OIDCService.CreateClientConfiguration()` for example

File: components/public-api/gitpod/experimental/v1/oidc.proto [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  // Optional.
+  string userinfo_id_key = 18;
+  // Optional.
+  string userinfo_name_key = 19;
+
+  // Based on this flag, the provider is not using OpenID for authentication,
+  // the request_type will be "code" instead of "id_token", and the
+  // "userinfo_endpoint" will be queried to obtain a profile by default.
+  bool is_oauth_only = 20;
+  // END – OAuth2 configuration
```
Comment (easyCZ) [inline components/public-api/gitpod/experimental/v1/oidc.proto:75]: Similarly here, we can define something like `OAuth2Spec` message and put the fields under there

**Comment by easyCZ** [comment]
LGTM. We'll likely iterate on this a bit before we get it just right but current form looks good!

**Review by easyCZ — APPROVED** [review]

**Pull request merged**